security/pkg/nodeagent/caclient/providers/vault/client.go

```diff
@@ -19,70 +19,135 @@ import (
 	"istio.io/pkg/log"
 )
 
+const (
+	certKeyInCACertResp        = "certificate"
+	certKeyInCertSignResp      = "certificate"
+	caChainKeyInCertSignResp   = "ca_chain"
+	issuingCAKeyInCertSignResp = "issuing_ca"
+	jwtKeyInLoginReq           = "jwt"
+	roleKeyInLoginReq          = "role"
+)
+
 var (
 	vaultClientLog = log.RegisterScope("vault", "Vault client debugging", 0)
 )
 
-type vaultClient struct {
-	enableTLS   bool
-	tlsRootCert []byte
+// VaultClient is a client for interaction with Vault.
+type VaultClient struct {
+	enableTLS       bool
+	vaultAddr       string
+	tlsRootCertPath string
+	jwtPath         string
+	loginRole       string
+	loginPath       string
+	signCsrPath     string
+	caCertPath      string
+
+	client    *vaultapi.Client
+	jwtLoader *util.JwtLoader
+}
 
-	vaultAddr        string
-	vaultLoginRole   string
-	vaultLoginPath   string
-	vaultSignCsrPath string
+// NewVaultClientWithConfig creates a CA client for the Vault PKI. It takes a single format config string.
+func NewVaultClientWithConfig(config string) (*VaultClient, error) {
+	params := strings.Split(config, ";")
+	if len(params) != 7 {
+		return nil, fmt.Errorf(
+			"error processing config for Vault. Expected 'backend_ca_addr;tls_cert_path;jwt_path;role;"+
+				"auth_path;sign_csr_path;ca_cert_path', but got '%s' which has %d segments", config, len(params))
+	}
+	vaultAddr := params[0]
+	tlsRootCertPath := params[1]
+	jwtPath := params[2]
+	vaultLoginRole := params[3]
+	vaultLoginPath := params[4]
+	vaultSignCsrPath := params[5]
+	vaultCACertPath := params[6]
 
-	client *api.Client
+	return NewVaultClient(
+		vaultAddr, tlsRootCertPath, jwtPath, vaultLoginRole, vaultLoginPath, vaultSignCsrPath, vaultCACertPath)
 }
 
-// NewVaultClient create a CA client for the Vault provider 1.
-func NewVaultClient(tls bool, tlsRootCert []byte,
-	vaultAddr, vaultLoginRole, vaultLoginPath, vaultSignCsrPath string) (caClientInterface.Client, error) {
-	c := &vaultClient{
-		enableTLS:        tls,
-		tlsRootCert:      tlsRootCert,
-		vaultAddr:        vaultAddr,
-		vaultLoginRole:   vaultLoginRole,
-		vaultLoginPath:   vaultLoginPath,
-		vaultSignCsrPath: vaultSignCsrPath,
+// NewVaultClient creates a CA client for the Vault PKI.
+func NewVaultClient(vaultAddr, tlsRootCertPath, jwtPath, loginRole, loginPath,
+	signCsrPath, caCertPath string) (*VaultClient, error) {
+	c := &VaultClient{
+		enableTLS:       true,
+		vaultAddr:       vaultAddr,
+		tlsRootCertPath: tlsRootCertPath,
+		jwtPath:         jwtPath,
+		loginRole:       loginRole,
+		loginPath:       loginPath,
+		signCsrPath:     signCsrPath,
+		caCertPath:      caCertPath,
+	}
+	if strings.HasPrefix(c.vaultAddr, "http:") {
+		c.enableTLS = false
 	}
 
-	var client *api.Client
+	jwtLoader, tlErr := util.NewJwtLoader(c.jwtPath)
+	if tlErr != nil {
+		return nil, fmt.Errorf("failed to create token loader to load the tokens: %v", tlErr)
+	}
+	stopCh := make(chan struct{})
+	go jwtLoader.Run(stopCh)
```

Comment:
"go jwtLoader.Run(stopCh)" indicates that the JWT reading is conducted asynchronously. Can you add a comment to the code about the lifecycle and state transitions of JWT reading?

Reply:
Sure, I'll update tomorrow. It's essentially a file watcher.